tests/ostree_mock.c

```diff
@@ -3,26 +3,135 @@
+ * [pacman].os == OSTREE_DEPLOYMENT_OS_NAME (aktualizr configuration, pacman section, os field must be equal to
+ * OSTREE_DEPLOYMENT_OS_NAME defined here)
+ *
+ * OSTREE_DEPLOYMENT_VERSION_FILE must follow this schema/format
+ *
+ * <deployed-revision-hash>\n
+ * <deployed-revision-serial>\n
+ * <booted-revision-hash>\n
+ * <booted-revision-serial>
+ *
+ * For example
+ *
+ *  d0ef921904bc573616a42e862de2421b265a76d06a334cfe5ad9f811107bbee0
+ *  0
+ *  d0ef921904bc573616a42e862de2421b265a76d06a334cfe5ad9f811107bbee0
+ *  0
+ */
+
+static const char OSTREE_DEPLOYMENT_OS_NAME[] = "dummy-os";
+static const char OSTREE_DEPLOYMENT_VERSION_FILE[] = "OSTREE_DEPLOYMENT_VERSION_FILE";
+
+static OstreeDeployment* ostree_sysroot_get_booted_deployment_from_file(const char* filename);
+
 OstreeDeployment* ostree_sysroot_get_booted_deployment(OstreeSysroot* self) {
-  OstreeDeployment* (*orig)(OstreeSysroot*) = dlsym(RTLD_NEXT, __func__);
-  char mocked_name[100];
-  snprintf(mocked_name, sizeof(mocked_name), "%s_mock", __func__);
-  OstreeDeployment* (*mocked)(OstreeSysroot*) = dlsym(RTLD_DEFAULT, mocked_name);
+  OstreeDeployment* (*orig)(OstreeSysroot*) = (OstreeDeployment * (*)(OstreeSysroot*))(dlsym(RTLD_NEXT, __func__));
+
+  const char* deployment_version_file = getenv(OSTREE_DEPLOYMENT_VERSION_FILE);
+  OstreeDeployment* deployment_from_file = NULL;
 
-  if (mocked) {
-    return mocked(self);
+  if (deployment_version_file != NULL &&
+      (deployment_from_file = ostree_sysroot_get_booted_deployment_from_file(deployment_version_file)) != NULL) {
+    return deployment_from_file;
   }
+
   return orig(self);
 }
 
-const char* ostree_deployment_get_bootcsum(OstreeDeployment* self) {
-  char* (*orig)(OstreeDeployment*) = dlsym(RTLD_NEXT, __func__);
-  char mocked_name[100];
-  snprintf(mocked_name, sizeof(mocked_name), "%s_mock", __func__);
-  char* (*mocked)(OstreeDeployment*) = dlsym(RTLD_DEFAULT, mocked_name);
+struct OstreeDeployment {
+  int serial;
+  char rev[255];
+};
+
+static int read_deployment_info(FILE* file, struct OstreeDeployment* deployment) {
+  struct OstreeDeployment deployment_res;
+
+  // TODO: make it more robust and reliable as a file might contain a value exceeding 255
+  if (fscanf(file, "%s\n", deployment_res.rev) != 1) {
+    return -1;
+  }
+
+  // TODO: make it more robust and reliable as a file might contain a value exceeding 255 and/or char(s) instead of a
+  // number for the serial value
+  if (fscanf(file, "%d\n", &deployment_res.serial) != 1) {
+    return -1;
+  }
+
+  *deployment = deployment_res;
+  return 0;
+}
+
+static int get_ostree_deployment_info(const char* filename, struct OstreeDeployment* deployed,
+                                      struct OstreeDeployment* booted) {
+  FILE* file = fopen(filename, "r");
+  if (file == NULL) {
+    printf("\n>>>>>>>>>> Failed to open the file: %s\n\n", filename);
+    return -1;
+  }
+
+  int return_code = -1;
+
+  while (1) {
```

Comment:
Why is this a loop? It looks like it cannot continue past the first iteration.

Reply:
To avoid `return` for each negative case.

Reply:
I see, then some `goto fail;` would be even more readable I think. This style calls for an explanatory comment, as it is quite uncommon. If you're really dogmatic against goto, it could be changed to a `do { } while (0);` which at least self-documents that this is not a loop. But it would probably still require a bit more time to grasp.

Reply:
I am fine with `do { } while (0);`. I've seen this approach on multiple projects (at least two) back in 20XX, perhaps I am getting too old  :)